[ALLUXIO-3192] Use recursive listStatus to implement du

yuzhu opened the pull request:
https://alluxio.atlassian.net/browse/ALLUXIO-3192

Use recursive listStatus to implement du

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19554/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='org.alluxio:alluxio-core-client-fs' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19554/org.alluxio$alluxio-core-client-fs/testReport'>org.alluxio:alluxio-core-client-fs</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19554/org.alluxio$alluxio-core-client-fs/testReport/alluxio.client.file.options/ListStatusOptionsTest/equalsTest/'><strong>alluxio.client.file.options.ListStatusOptionsTest.equalsTest</strong></a></li></ul><h3><a name='org.alluxio:alluxio-core-server-master' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19554/org.alluxio$alluxio-core-server-master/testReport'>org.alluxio:alluxio-core-server-master</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19554/org.alluxio$alluxio-core-server-master/testReport/alluxio.master.file/FileSystemMasterTest/listStatus/'><strong>alluxio.master.file.FileSystemMasterTest.listStatus</strong></a></li><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19554/org.alluxio$alluxio-core-server-master/testReport/alluxio.master.file.options/ListStatusOptionsTest/equalsTest/'><strong>alluxio.master.file.options.ListStatusOptionsTest.equalsTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19556/
Test PASSed.

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+    try (InodeLockList lockList = new InodeLockList()) {
+      while (!queue.isEmpty()) {
+        LockedInodePath currInodePath = queue.remove();
+        Inode<?> inode = currInodePath.getInode();
```
this only works for a full path. Are these guaranteed to be full? Are we expecting that the target inode always exists?

yuzhu replied:
For the `TempInodePathForDescendant` kind, this does not really apply. For the very first node that we deal with, it is guaranteed to be full. since we call `ensureFullPathAndUpdateCache` before calling `listStatusInternal`, but maybe we should check again?

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-          } finally {
-            child.unlockRead();
+            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, originalInodePath);
+          } catch (Exception e) {
```
Should we catch `AccessControlException`?

yuzhu replied:
done

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-            child.unlockRead();
+            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, originalInodePath);
+          } catch (Exception e) {
+            auditContext.setAllowed(false);
```
How does this work on a local FS? If you do a `ls -R`, and you don't access to some dirs, does the ls fail, or do you see the ones you do have access to?

yuzhu replied:
I tried it on a local FS. it fails for some and returns the result for others that you do have access to.  I can return result for those ones that you do have access to, but will silently fail for those that you don't have access to.

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+              TempInodePathForDescendant tempInodePath
+                  = new TempInodePathForDescendant(originalInodePath);
+              tempInodePath.setDescendant(child, mInodeTree.getPath(child));
+              queue.add(tempInodePath);
```
What do you think if we depth-first (lock child, recursive call, unlock) instead of breadth-first? I am asking because we may be able to unlock descendants earlier, which can allow more concurrency?

I think the output might be more meaningful too?

yuzhu replied:
done

gpang commented on core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -908,6 +908,26 @@ public void listStatus() throws Exception {
           filenames.contains(NESTED_URI.join("file" + String.format("%05d", i)).toString()));
     }
 
+    // Test recursive listStatus
```
Can you put this in its own test case?

yuzhu replied:
done.

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-        }
-        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
+      listStatusRecursive(ret, inodePath, auditContext, listStatusOptions.isRecursive());
```
done

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+    }
+  }
+
+  private void listStatusRecursive(List<FileInfo> statusList, LockedInodePath originalInodePath,
```
done

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19562/<h2>Build result: FAILURE</span></h2>[...truncated 2069 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19564/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-core-server-master' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19564/org.alluxio$alluxio-core-server-master/testReport'>org.alluxio:alluxio-core-server-master</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19564/org.alluxio$alluxio-core-server-master/testReport/alluxio.master.file/PermissionCheckTest/readNotExecuteDir/'><strong>alluxio.master.file.PermissionCheckTest.readNotExecuteDir</strong></a></li></ul>
Test FAILed.

yuzhu commented:
@gpang This failed test is due to us not throwing an exception when we don't have permission for a particular directory.  We could throw the exception when recursive is turned off, and not throw it when recursive is turned on?   Addressed other comments, PTAL.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19565/<h2>Build result: FAILURE</span></h2>[...truncated 2067 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19567/
Test PASSed.

gpang commented on core/server/master/src/main/java/alluxio/master/file/meta/InodeLockList.java:
```diff
@@ -26,7 +26,10 @@
+  /**
+   * Creates a new instance of {@link InodeLockList}.
+   */
+  public InodeLockList() {
```
is this change no longer necessary?

yuzhu replied:
removed

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-            ret.add(getFileInfoInternal(tempInodePath));
-          } finally {
-            child.unlockRead();
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
```
hrmmm, is this because we don't have a temp way to extend a lock list?

yuzhu replied:
Yes. For reentrant locks, this is fine though, since we will never upgrade a lock from read to write?

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-      return ret;
     }
+    // We add the file info of all files and any directory that is not the original directory
+    if (currInodePath.getInode().isFile() || !currInodePath.equals(originalInodePath)) {
```
why are only files added?

Also, can we just add all the paths to the list? If we do a preorder traversal, we would always add the input path first. If we don't want the first input path, we can detect that with an empty list.

yuzhu replied:
Files and directories are added. The only exception is when the original inode is a directory, then we don't add it to the list.  I am not sure how we can throw away the first input path with an empty list?

gpang replied:
Well, if we always insert the input inode, and then its children, when the output list is empty, we know the first inode we can insert is the very first input inode. Therefore, we can see if we want to insert that one or not, without having to pass in `originalInodePath`.

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+                                                  boolean recursive,
+                                                  LockedInodePath originalInodePath)
+      throws FileDoesNotExistException, UnavailableException, AccessControlException {
+    List<FileInfo> statusList = new ArrayList<>();
```
I see. Since now `listStatusInternal` is recursive, maybe we should pass in the mutable list so the recursion can just add to a single list. Constructing many lists sounds bad. What do you think?

yuzhu replied:
Yeah, I agree.

gpang commented:
also, can you resolve the conflicts?

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -929,32 +938,51 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
+      List<FileInfo> ret = listStatusInternal(inodePath, auditContext,
+          listStatusOptions.isRecursive(), inodePath);
```
I passed the original inode into the function to avoid adding the original inode to the list if it is a directory.

madanadit commented on core/client/fs/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -90,7 +109,8 @@ public boolean equals(Object o) {
     return Objects.equal(mCommonOptions, that.mCommonOptions)
-        && Objects.equal(mLoadMetadataType, that.mLoadMetadataType);
+        && Objects.equal(mLoadMetadataType, that.mLoadMetadataType)
+        && Objects.equal(mRecursive, that.mRecursive);
```
add to hashcode and toString as well

yuzhu replied:
done

madanadit commented on core/server/master/src/main/java/alluxio/master/file/options/ListStatusOptions.java:
```diff
@@ -104,7 +125,8 @@ public boolean equals(Object o) {
     return Objects.equal(mLoadMetadataType, that.mLoadMetadataType)
-        && Objects.equal(mCommonOptions, that.mCommonOptions);
+        && Objects.equal(mCommonOptions, that.mCommonOptions)
+        && Objects.equal(mRecursive, that.mRecursive);
```
same here. add to hashcode and toString

yuzhu replied:
done

madanadit commented on core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -918,6 +918,41 @@ public void listStatus() throws Exception {
+    for (int i = 0; i < files; i++) {
+      createFileWithSingleBlock(ROOT_URI.join("file" + String.format("%05d", i)));
+    }
+    // Test files in nested directory.
```
does this create multi-level nested directories?

yuzhu replied:
yes

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19595/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.alluxio:alluxio-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19595/org.alluxio$alluxio-tests/testReport'>org.alluxio:alluxio-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Alluxio-Pull-Request-Builder/19595/org.alluxio$alluxio-tests/testReport/alluxio.client.fs/ConcurrentFileSystemMasterCreateTest/concurrentListFiles/'><strong>alluxio.client.fs.ConcurrentFileSystemMasterCreateTest.concurrentListFiles</strong></a></li></ul>
Test FAILed.

yuzhu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19597/
Test PASSed.

yuzhu commented:
@gpang  @madanadit Addressed your comments PTAL. Unfortunately the review history got clobbered because I rebased on the tip of the master to resolve the conflict. Let me know if there is a better way to address conflicts.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19598/
Test PASSed.

gpang commented:
@yuzhu For resolving conflicts, you can just merge master into your branch, resolve conflicts, and commit that merge commit. That should preserve all the history.

gpang commented on core/client/fs/src/main/java/alluxio/client/file/options/ListStatusOptions.java:
```diff
@@ -90,19 +109,21 @@ public boolean equals(Object o) {
     return Objects.toStringHelper(this)
         .add("commonOptions", mCommonOptions)
         .add("loadMetadataType", mLoadMetadataType.toString())
+        .add("Recursive", mRecursive)
```
NIT: "recursive"

gpang commented on core/server/master/src/main/java/alluxio/master/file/options/ListStatusOptions.java:
```diff
@@ -104,19 +125,21 @@ public boolean equals(Object o) {
     return Objects.toStringHelper(this)
         .add("commonOptions", mCommonOptions)
         .add("loadMetadataType", mLoadMetadataType.toString())
+        .add("Recursive", mRecursive)
```
NIT: "recursive"

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
     }
   }
 
+  private void listStatusInternal(LockedInodePath currInodePath,
```
can you javadoc this method, especially since the output is passed in via `statusList`, and what order this is being returned?

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+
+      // If we are listing the status of a directory, we remove the directory info that we inserted
+      if (inode.isDirectory() && ret.size() >= 1) {
+        ret.remove(ret.size() - 1);
```
is this expensive?

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+          throw e;
+        }
+      }
+      if (descendantType == DescendantType.ALL || descendantType == DescendantType.ONE) {
```
`descendantType != NONE`?

yuzhu replied:
we don't need to do anything when `descendantType != NONE`

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+        DescendantType nextDescendantType = (descendantType == DescendantType.ALL)
+            ? DescendantType.ALL : DescendantType.NONE;
+        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
```
We should add a TODO to lock this more efficiently.

yuzhu replied:
done

gpang commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,58 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+          try (LockedInodePath childInodePath = mInodeTree.lockFullInodePath(child.getId(),
+              InodeTree.LockMode.READ)) {
+            listStatusInternal(childInodePath, auditContext,
+                nextDescendantType, statusList);
```
NIT: will this fit on the previous line?

yuzhu replied:
done

yuzhu commented:
@gpang addressed comments, PTAL. Thanks!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19618/
Test PASSed.

madanadit commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -909,7 +917,8 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
             && listStatusOptions.getLoadMetadataType() != LoadMetadataType.Always
-            && ((InodeDirectory) inode).isDirectChildrenLoaded()) {
+            && ((InodeDirectory) inode).isDirectChildrenLoaded()
+            && !listStatusOptions.isRecursive()) {
```
will a recursive listing always load from the ufs even for `LoadMetadataype.Once`?

gpang replied:
Maybe this is the same question, but how does `recursive` interact with load metadata?

yuzhu replied:
I initially thought `recursive` would force us to load because our only way to detect if a subtree is loaded once is by calling `isDirectChildrenLoaded`, and that does not really tell us about grandchildrens etc. but if there is a way to respect `loadMetadata` flag, specifically `LoadMetadataype.Once`, I think that would be good..

madanadit replied:
sounds like the `loaded` bit needs to be modified so that it can give your more info: `never loaded`, `direct children loaded` or `recursively loaded`

madanadit replied:
@gpang I was talking about inode.mDirectChildrenLoaded which is a single bit. note: "loaded" vs load behavior. makes sense?

yuzhu replied:
addressed as discussed.

madanadit commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +929,72 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-        }
-      } else {
-        ret.add(getFileInfoInternal(inodePath));
+      DescendantType descendantTypeForListStatus =  (listStatusOptions.isRecursive())
```
extra space after `=`

yuzhu replied:
done.

madanadit commented on core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -918,6 +918,41 @@ public void listStatus() throws Exception {
   }
 
+  @Test
+  public void listStatusRecursive() throws Exception {
```
is it easy to add tests with `Permissions`? test recursive listing honors permissions for nested directories

yuzhu replied:
done

yuzhu commented:
addressed the comments, PTAL! @gpang @madanadit

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19647/<h2>Build result: FAILURE</span></h2>[...truncated 2068 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

yuzhu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19648/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19649/
Test PASSed.

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectory.java:
```diff
@@ -171,6 +171,24 @@ public synchronized boolean isDirectChildrenLoaded() {
+  /**
+   * @return true if we have loaded all the direct and indirect children's metadata once
+   */
+  public synchronized boolean isDescendantLoaded() {
```
done

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +935,72 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+            listStatusInternal(childInodePath, auditContext,
+                nextDescendantType, statusList);
+          } catch (Exception e) {
+            LOG.debug("ListStatus failed for path {}", mInodeTree.getPath(child).getPath(), e);
```
similar reason, if we failed to list status for one of the subdirectories, should we at least return some results?  I figured log and return some results is the best we can do.

madanadit replied:
`Exception` is to general is masking a lot of types. I would rather you catch different access types and report to the consumer

yuzhu replied:
done

yuzhu commented on core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -918,6 +919,53 @@ public void listStatus() throws Exception {
   }
 
+  @Test
+  public void listStatusRecursive() throws Exception {
```
done

yuzhu commented on core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java:
```diff
@@ -918,6 +919,53 @@ public void listStatus() throws Exception {
+          filenames.contains(NESTED_URI.join("file" + String.format("%05d", i)).toString()));
+    }
+
+    // Test with permissions
```
done

yuzhu commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -891,26 +891,41 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
-            && ((InodeDirectory) inode).isDirectChildrenLoaded()) {
-          loadMetadataOptions.setLoadDescendantType(DescendantType.NONE);
+            && listStatusOptions.getLoadMetadataType() != LoadMetadataType.Always) {
+          if (listStatusOptions.isRecursive()
```
done

yuzhu commented:
@gpang @madanadit addressed comments, PTAL. Thanks!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19691/<h2>Build result: FAILURE</span></h2>[...truncated 2078 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder@2/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19692/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19699/
Test PASSed.

madanadit commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -920,36 +935,72 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
+        mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, currInodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        if (descendantType == DescendantType.ALL) {
```
The posix behavior shows you the entries you have access to and reports a permission denied error message for what you don't. I think we should do something similar instead of silently ignoring

gpang replied:
I'm not sure what the best way to do this is, since it looks like we want to return results back, but also return exception/error messages. I don't think we have a mechanism to do both.

yuzhu replied:
maybe we can create a subclass of `FileInfo` called `ErrorFileInfo` that wraps the exception? and return that?

yuzhu replied:
Added a TODO for this item

gpang commented on core/server/master/src/main/java/alluxio/master/file/PermissionChecker.java:
```diff
@@ -43,8 +43,9 @@ void checkParentPermission(Mode.Bits bits, LockedInodePath inodePath)
    * @throws AccessControlException if permission checking fails
    * @throws InvalidPathException if the path is invalid
    */
+
```
Remove this extra line?

yuzhu replied:
done

yuzhu commented:
@gpang @madanadit addressed your comment. added TODO for the exception handling case. PTAL.  Thanks.

yuzhu commented:
jenkins, test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19711/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19712/
Test PASSed.
